### PR TITLE
py/formatfloat: Calculate reference values with pow().

### PR DESCRIPTION
**Pull request by dpwe**
Formerly, ```formatfloat``` struggled with formatting exact powers of 10
as created by ```parsenum``` because of small differences in how those
powers of 10 were calculated: ```formatfloat``` multiplied together a few
values of 10^(2^Y) from a lookup table, but ```parsenum``` used
```pow(10, X)```.  This was mitigated in #8905 by adding 2eps of extra
margin when comparing values ("aggressive rounding up").

This patch instead eliminates the discrepency by using ```pow()``` in
```formatfloat``` also.  This eliminates the need for the 2eps margin, as
well as the lookup tables.  It is likely slower to run, however.

In addition, this patch directly estimates the power-of-10 exponent from
the power-of-2 exponent in the floating-point representation.

This change surfaced a previously-noted problem with parsing numbers
including trailing zeros.  The fix to that problem, #8980, is included
in this PR to ensure the tests pass.

Signed-off-by: Dan Ellis <dan.ellis@gmail.com>

### CODE REVIEW

**Comment by dpwe** [comment]
@dpgeorge Any hints as to why this is failing on appveyor?  All I can see is the build error outputs, which seem catastrophic - my code prints nothing for numbers smaller than 1.  But I can't see how this is happening (and can't reproduce, obviously). The Appveyor build appears to be under Windows using an MS compiler - is there some reason why mp_float_union_t wouldn't work there?  I got rid of the special case float_union_t (for accessing the bits in a 32-bit float) from formatfloat since it appeared redundant with mp_float_union_t, but now I'm doubting my actions.  In particular, the new code computes directly from the exp bits extracted from mp_float_union_t, maybe these read as signed under the MS compiler or something?

**Comment by dpwe** [comment]
Although the tests pass, there's a big problem with this.  Using a float32 build, 
```
>>> print("%.8e" % 1e-13)
0.99999990e-13
```
That's a malformed "e"-format number, the leading digit cannot be 0 in "e"-format.

The problem occurs with a mismatch between how the leading digit place is calculated, and how the leading digit itself is calculated.  For numbers smaller than 1.0, we decide the leading digit by comparing the input number with ```pow(10, -e)``` for increasing values of ```e``` until we get something less than or equal to the input number. In the case of ```1e-13```, that returns ```e = 13``` as expected.  

We then scale up the input by ```pow(10, e)```, and the leading digit is however many times we can subtract 1 from this value and not go negative.  Unfortunately, for some values of ```e``` (specifically, 13, 17, 19, 28, 31, and 38), the product ```pow(10, e) * pow(10, -e)``` is smaller than 1, e.g.:
```
>>> print("%.12f" % ((1e13) * (1e-13)))
0.999999904633
```
Thus, when we come to see how many units we can fit in the scaled-up input, it's zero, and we print the incorrectly-normalized "e"-format number above.

I think this was being hidden by the hard ```FPROUND_TO_ONE``` logic that I didn't understand and removed in this PR, but I'm looking for a cleaner solution.

**Comment by dpwe** [comment]
This code is more understandable, I believe more accurate (no aggressive rounding), and should even be smaller (although the changes to ```parsenum``` will have increase code size there).  PTAL.

**Comment by dpgeorge** [comment]
Code size diff for this PR (which includes #8980):
```
   bare-arm:    +0 +0.000% 
minimal x86:    +0 +0.000% 
   unix x64:  -192 -0.031% 
unix nanbox:  -208 -0.037% 
      stm32:   -52 -0.013% PYBV10
     cc3200:    -8 -0.004% 
    esp8266:   -48 -0.007% GENERIC
      esp32:  +168 +0.011% GENERIC[incl -48(data)]
     mimxrt:  -184 -0.052% TEENSY40
 renesas-ra:   -48 -0.008% RA6M2_EK
        nrf:   -32 -0.017% pca10040
        rp2:  -176 -0.034% PICO
       samd:   -52 -0.037% ADAFRUIT_ITSYBITSY_M4_EXPRESS
```
That's great!

File: py/formatfloat.c [hunk 1]
```diff
@@ -167,7 +144,6 @@ int mp_format_float(FPTYPE f, char *buf, size_t buf_size, char fmt, int prec, ch
             return s - buf;
         }
     }
-
```
Comment (dpgeorge) [inline py/formatfloat.c:170]: please don't add/remove unrelated blank lines unless it makes sense to the flow of the code that's being changed

File: py/formatfloat.c [hunk 2]
```diff
@@ -262,19 +233,11 @@ int mp_format_float(FPTYPE f, char *buf, size_t buf_size, char fmt, int prec, ch
...
-            // rendering as 9.99999998e(X-1).
-            if (fp_ge_eps(f, next_u)) {
-                u_base = next_u;
-                e += e1;
-            }
+        e = e_guess;
+        FPTYPE next_u = MICROPY_FLOAT_C_FUN(pow)(10, e + 1);
+        while (f >= next_u) {
+            ++e;
+            next_u = MICROPY_FLOAT_C_FUN(pow)(10, e + 1);
```
Comment (dpgeorge) [inline py/formatfloat.c:240]: this loop could potentially be very slow, eg for a large number like 1e300 it would run 300 times??

File: py/parsenum.c [hunk 3]
```diff
@@ -311,7 +340,7 @@ mp_obj_t mp_parse_num_float(const char *str, size_t len, bool allow_imag, mp_lex
         }
 
         // apply the exponent, making sure it's not a subnormal value
-        exp_val += exp_extra;
+        exp_val += (exp_extra + trailing_zeros_intg);
```
Comment (dpgeorge) [inline py/parsenum.c:343]: no need to add parenthesis here

**Comment by dpgeorge** [comment]
This looks like a good improvement, because it reduces code size and fixes the trailing 0 problem.  And as you mention it no longer does the aggressive rounding up.

Can you please rebase this on master and clean up the commits, with at least a separate commit to fix the trailing zero problem, and then 1 or more extra commits for the other changes.  (At the moment the commits conflict, it seems it was merged to itself.)

File: py/formatfloat.c (hunk 2 above)
Comment (dpwe) [reply py/formatfloat.c:240]: Fortunately, the loop starts from e_guess which is calculated from the binary exponent in the float, so it only has to try one or two values of e (i.e., e_guess is at worse the final target e minus one).

File: py/formatfloat.c (hunk 1 above)
Comment (dpwe) [reply py/formatfloat.c:170]: fixed

File: py/parsenum.c (hunk 3 above)
Comment (dpwe) [reply py/parsenum.c:343]: fixed

**Comment by dpwe** [comment]
Sounds good.  I have rebased and cleaned up the commit history as you requested.  I have rebased to HEAD, then there are now two commits, one adds the changes to parsenum.c, and the second adds the changes to formatfloat.c and the tests.  

I'm getting more successful with ```git rebase```; however, there are a couple of formatting changes to ```parsenum``` in the second commit that don't belong there.  I'm hoping this is OK.

File: py/formatfloat.c (hunk 2 above)
Comment (dpgeorge) [reply py/formatfloat.c:241]: great!

**Comment by dpgeorge** [comment]
Thanks for updating the commits.

Rebased and merged in 6cd2e4191803e95580bdfc57c06ea818454a25d1 and 6f4d424f461bede1afb69637763f4fce5f27cd90

-------

Final code size diff for this PR is:
```
   bare-arm:    +0 +0.000% 
minimal x86:    +0 +0.000% 
   unix x64:  -144 -0.023% standard
unix nanbox:  -208 -0.037% nanbox
      stm32:   -44 -0.011% PYBV10
     cc3200:    +0 +0.000% 
    esp8266:   -44 -0.006% GENERIC
      esp32:  +156 +0.010% GENERIC[incl -48(data)]
     mimxrt:  -176 -0.049% TEENSY40
 renesas-ra:   -48 -0.008% RA6M2_EK
        nrf:   -24 -0.013% pca10040
        rp2:  -152 -0.030% PICO
       samd:   -44 -0.031% ADAFRUIT_ITSYBITSY_M4_EXPRESS
```
